Upgrade Antlr3 and 4. New toml parsing library.

vieiro opened the pull request:
Hi all,

As suggested in PR https://github.com/apache/netbeans/pull/6844 we should upgrade Antlr3 and Antlr4 runtimes to latest versions (3.5.3 and 4.13.1).

Problem is that `ide/libs.tomlj` is still using Antlr4.11.1, and also has [(problems parsing some TOML files](https://github.com/tomlj/tomlj/issues/66) and [is not actively maintained](https://github.com/tomlj/tomlj/issues/44).

This PR

- Upgrades Antlr3 and Antlr4 runtimes to latests versions (3.5.3, 4.13.1)
- Adds `ide/tomljava` as a new wrapper module around [toml-java](https://github.com/vieiro/toml-java)
- Keeps the existing `ide/tomlj`, that should be marked as deprecated (how is this done?) and possibly removed in the future.
- Modifies `ide/languages.toml` to use the new library, and adds semantic coloring and structure in the navigator (screenshot below).


![image](https://github.com/apache/netbeans/assets/564140/767c9f6f-861d-4568-9130-44467f2e3742)

vieiro commented:
It seems our `ValidateClassFilesTest` is complaining about `libs/tomljava` not being marked as Java 11. Will fix as time permits.

mbien commented on ide/languages.toml/nbproject/project.properties:
```diff
@@ -14,6 +14,6 @@
 # specific language governing permissions and limitations
 # under the License.
-javac.source=1.8
+javac.source=11
```
for consistency reasons please set source **and** target:
```
javac.source=11
javac.target=11
```
at some point we will search/replace over that and use only `javac.release=11` where possible.

mbien commented:
added all-tests label just to be sure. If it was green once, feel free to remove it again.

vieiro commented:
Yep, I've also added two more images to `licenseinfo.xml`, and updated `javac.target` in antlr4 runtime lib (antlr4.13.1 is compiled with JDK11). Let's see how it goes with full tests now.

mbien commented:
all green!

mbien commented:
> Keeps the existing `ide/tomlj`, that should be marked as deprecated (how is this done?) and possibly removed in the future.

I am not sure but probably over the module manifest, example:
https://github.com/apache/netbeans/blob/306f174318db1e1d5fbe7a63d827d66c7a38978d/ide/o.openidex.util/manifest.mf#L5-L6

I believe this will cause a warning during module load.

vieiro commented:
So let's deprectate `libs.tomlj` then. It's currently used in the `rust` cluster only and will be replaced later on (and possibly by platform users).

mbien commented on ide/libs.tomljava/manifest.mf:
```diff
@@ -0,0 +1,4 @@
+Manifest-Version: 1.0
+OpenIDE-Module: org.netbeans.libs.tomljava/3
+OpenIDE-Module-Localizing-Bundle: org/netbeans/libs/tomljava/Bundle.properties
+OpenIDE-Module-Specification-Version: 1.0
```
consider adding:
```
OpenIDE-Module-Java-Dependencies: Java > 11
```
This would prevent the module and the dependency tree to be loaded on platform apps using older JDKs. Otherwise it will fail later. (I don't think the module system looks at source/target)

vieiro replied:
So that should say `OpenIDE-Module-Java-Depencencies: Java >= 11`, right? Worth a change?

mbien replied:
its confusing but `>` means `>=` 

we have this everywhere https://github.com/search?q=repo%3Aapache%2Fnetbeans%20OpenIDE-Module-Java-Dependencies%3A%20Java%20%3E%2011&type=code

neilcsmith-net replied:
It also should be failing a test if it's not there?!

mbien replied:
not sure if the test takes lib wrappers into account, it might only look at class files of the project file?

vieiro replied:
AFAIU library wrapper class versions are not taken into account in the current `ant commit-validation` tests.

neilcsmith-net replied:
No, sorry, misread which manifest this was.

neilcsmith-net replied:
Actually, looking back at #4775 and my change before that, I think libraries are meant to be checked too. It used to be every JAR. Is this definitely compiled for release 11?

vieiro replied:
antlr4-runtime-4.13.1 is still Java 8 (thought it was 11), toml-java is Java 11.

```sh
$ javap -v -cp antlr4-runtime-4.13.1.jar org.antlr.v4.runtime.CharStream
Classfile jar:file:///[...]/netbeans/ide/libs.antlr4.runtime/external/antlr4-runtime-4.13.1.jar!/org/antlr/v4/runtime/CharStream.class
  Last modified 4 sept 2023; size 237 bytes
  SHA-256 checksum f56df94dc6d8100a59ee452da9e9cdd30b94523ce8e297daeab781193aed81c1
  Compiled from "CharStream.java"
public interface org.antlr.v4.runtime.CharStream extends org.antlr.v4.runtime.IntStream
  minor version: 0
  major version: 52
```

neilcsmith-net replied:
OK, the test should be picking up a problem here as far as I can tell.  I think it might have become too lenient!  I think there's an issue stemming from #4775 where it's ignoring modules without a declared Java dependency, and I think also ignoring non-referenced autoloads.  The test looks like it potentially needs `honorAutoloadEager` switched off, and to always add a default version for all module referenced JARs in https://github.com/apache/netbeans/blob/master/platform/o.n.core/test/qa-functional/src/org/netbeans/core/validation/ValidateClassFilesTest.java#L129

Interestingly, also noticed that there is code to add Java dependency automatically for value of `javac.target` that has a now incorrect regex protecting it?? https://github.com/apache/netbeans/blob/master/nbbuild/antsrc/org/netbeans/nbbuild/JarWithModuleAttributes.java#L221  (I was looking for where this dependency was being added automatically when I noticed the test now relies on its presence).

mbien replied:
regarding JarWithModuleAttributes.java: we should check `javac.release` there too. However the regexp looks good to me, no? I didn't know that this existed - this essentially makes all the manually added manifest fields redundant, no?

neilcsmith-net replied:
I assumed from the conversation above that the built module manifest didn't have the version in it.  Checking the dev build shows this attribute is included already.  I misread the regexp as if using `+` rather than `*` because I thought it wasn't passing the if check. All in all a long way around to no change required here then? :smile:

The test does still need looking at, but should be passing for the right reasons here.

And, yes, explicitly added manifest fields should be redundant for cases where `javac.target` is set then.  There are a few places this is used where this wasn't the case (eg. JShell support IIRC).

vieiro replied:
In fact I can easily downgrade `toml-java` to JDK8 binary format for the time being. Let's do that.

mbien replied:
@vieiro no wait, no downgrade is needed - (almost) everything is working as intended.

vieiro replied:
D'oh! I've downgraded to JDK8. No worries, we can always upgrade to JDK11.
After all the objective of `toml-java` was for it to be reusable on other projects, the lower the JDK the more projects can use it.

mbien replied:
@neilcsmith-net scraped a PR together from the things we discovered in this discussion: https://github.com/apache/netbeans/pull/7188

mbien reviewed (approved):
code looks good to me. I only quickly tested a few tomls and it worked great (haven't tested any other antlr based editors though).

thanks for working on this!

lkishalmi commented:
There are a few places in the IDE where the ANTLR generated java files are in the codebase. These are not re-generated during build time. and would cause warnings on the output:
```
ANTLR Tool version 4.11.1 used for code generation does not match the current runtime version 4.13.1
```

Fortunately it's nothing more serious than that this time.

Affected languages, that I remember: ANTLR v3, ANTLR v4, Go, JSON, maybe CSS

mbien commented:
good that you mentioned this since this reminded me of the note on the antlr4 page:
 >- minor version updates may include minor breaking changes, the policy is to regenerate parsers with every release (4.11 -> 4.12)
> - backwards compatibility is only guaranteed for patch version bumps (4.11.1 -> 4.11.2)

so the antlr3 (3.5.2->3.5.3) update was a patch update and should be fine, however the antlr4 (4.11.1->4.13.1) update might need regen.

lkishalmi commented:
Referencing the previous upgrade from 4.7.2 -> 4.11.1 #4845

vieiro commented on ide/languages.toml/src/org/netbeans/modules/languages/toml/semantic/TomlSemanticAnalyzer.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public void run(TomlParser.TomlParserResult result, SchedulerEvent event) {
+        synchronized (this) {
+            if (cancelled) {
+                return;
+            }
+        }
```
I think this is more understandable: an instance variable being modified synchronously. `AtomicBoolean` would also do (`volatile` is not required). I think it's good as it is.

mbien replied:
yeah, its not a big deal here but volatile / AtomicBoolean would be much more performant compared to critical secitons. It is also a common idiom in NB to use AtomicBoolean for cancellation checks - you can see this often in editor code:
https://github.com/search?q=repo%3Aapache%2Fnetbeans+%22atomicboolean+cancelled%22&type=code

vieiro replied:
Done.

matthiasblaesing commented:
> so the antlr3 (3.5.2->3.5.3) update was a patch update and should be fine, however the antlr4 (4.11.1->4.13.1) update might need regen.

So every minor update is an API break? This disqualified antlr outside a very narrow of project. Everywhere where multiple dependencies might rely on antlr is thus brittle. Or am I missing something?

vieiro commented:
> There are a few places in the IDE where the ANTLR generated java files are in the codebase. These are not re-generated during build time. and would cause warnings on the output:
> 
> ```
> ANTLR Tool version 4.11.1 used for code generation does not match the current runtime version 4.13.1
> ```
> 
> Fortunately it's nothing more serious than that this time.
> 
> Affected languages, that I remember: ANTLR v3, ANTLR v4, Go, JSON, maybe CSS

Good point! Anyway in order to avoid problems in the future these grammars should be rebuilt with the IDE's Antlr runtime, right?

- `webcommon/javascript2json` [currently rebuilds the grammar](https://github.com/apache/netbeans/blob/a74ce3b45ce698269109b87c984ca6d8d96b4471/webcommon/javascript2.json/build.xml#L29) so we should be safe.
- `ide/languages.toml` currently doesn't.
- `ide/languages.go` [currently doesn't, there's no grammar there](https://github.com/apache/netbeans/blob/a74ce3b45ce698269109b87c984ca6d8d96b4471/ide/languages.go/build.xml#L27), so we can't regenerate the lexer/parser with new runtime.
- `ide/lexer.antlr4` builds a grammar for testing purposes, but [there's no lexer/grammar for the Antlr4 lexer either](https://github.com/apache/netbeans/blob/a74ce3b45ce698269109b87c984ca6d8d96b4471/ide/lexer.antlr4/build.xml#L32)

I think we should add the .g4 lexers and grammars to the source code, and rebuild them with IDE's runtime, right?

neilcsmith-net commented on ide/libs.tomljava/external/toml-java-13.3.1-license.txt:
```diff
@@ -0,0 +1,208 @@
+Version: 13.3.1
+Description: A parser for Tom's Obvious, Minimal Language (TOML).
+License: Apache-2.0
+Origin: Apache Software Foundation
```
This needs changing to reference you and not the ASF?

mbien replied:
Antonios's parser of Tom's Obvious, Minimal Language: ATOML

neilcsmith-net replied:
Sure. as long as it doesn't say `Origin: Apache Software Foundation`. :smile:

vieiro replied:
Done.

mbien commented:
> I was wondering if we should also bump the specification version

@vieiro the way this works is that the spec version of all modules is bumped once per release in one big PR (last: https://github.com/apache/netbeans/pull/6959). However, some lib wrapper modules tend to have a similar version number than the lib they wrap. So bumping those IMO is not needed unless:

- you like to bump them for clarity/cosmetic reasons
- you want to have the dependency requirement (project.xml) of another module updated at the same time as the lib wrapper and don't want to wait for the global dep bump

neilcsmith-net reviewed (commented):
I think the library origin reference needs checking.

Looks like addition of `javac.target` got the test to pass by removing the module Java version dependency.  We might want to fix the test and/or the addition of the dependency in this or before this merges?

neilcsmith-net commented:
> However, some lib wrapper modules tend to have a similar version number than the lib they wrap

Or take the approach of the FlatLaf wrapper and use the upstream version as implementation version - https://github.com/apache/netbeans/blob/master/platform/libs.flatlaf/manifest.mf#L7

vieiro commented:
Rebasing on top of #7189 and #7186 for a final test.
If tests pass and nobody opposes this will be merged within in a couple days.

lkishalmi reviewed (approved):
I think it's good to go.

The build for `webcommon/javascript2.json` still have to be altered to generate the sources on the fly though.

I'm traveling today. So a PR for that can be expected from me a bit later.

lkishalmi reviewed (commented):
Well I was thinking of removing the generated code from the code base. And have the sources generated build time.

This is fine, just have to be repeated each ANTLR tool update...

vieiro commented:
> Well I was thinking of removing the generated code from the code base. And have the sources generated build time.
> 
> This is fine, just have to be repeated each ANTLR tool update...

Ah, I see. Let's give it a run then. Feel free to cherry-pick last commit and modify at will.

It seems the parser is a part of the API (used in three other modules, so let me double-check again we don't break compatibility).

lkishalmi commented:
Well I'd keep the `JsonBaseListener` and `JsonBaseVisitor` interfaces, mark them `@Deprecated`, maybe even for removal. Though right now they are just a compatibility layer of the ANTLR4 generated interfaces. (See: https://github.com/apache/netbeans/pull/4845#issuecomment-1288019365)

vieiro commented:
> Well I'd keep the `JsonBaseListener` and `JsonBaseVisitor` interfaces, mark them `@Deprecated`, maybe even for removal. Though right now they are just a compatibility layer of the ANTLR4 generated interfaces. (See: [#4845 (comment)](https://github.com/apache/netbeans/pull/4845#issuecomment-1288019365))

Well, the API is exported to friend modules only (currently 3: `o.n.m.javascript2.[editor,lexer,model]`) and these have been already upgraded, so I don't see a reason to keep 4.11.1 generated code around. But hey, if you want to keep these for another release then that's OK for me. I'll modify the last commit in a few hours. Also I'll fire a new ci-all-tests CI check.
